host/xtest/asym_perf.c

```diff
@@ -0,0 +1,962 @@
+					 TEEC_MEMREF_PARTIAL_OUTPUT,
+					 TEEC_VALUE_INPUT,
+					 TEEC_MEMREF_TEMP_INPUT);
+	op.params[0].memref.parent = (mode == MODE_DECRYPT) ?
+				     &in_shm : &hash_shm;
+	op.params[0].memref.size = (mode == MODE_DECRYPT) ?
+				   size : hash_shm.size;
+	op.params[1].memref.parent = &out_shm;
+	op.params[1].memref.size = out_shm.size;
+	op.params[2].value.a = mode;
+	op.params[3].tmpref.buffer = buf;
+	op.params[3].tmpref.size = blen;
+
+	res = TEEC_InvokeCommand(&sess, cmd, &op, &ret_origin);
+	check_res(res, "TEEC_InvokeCommand()", &ret_origin);
+	out_shm.size = op.params[1].memref.size;
+}
+
+static void do_warmup(int warmup)
+{
+	struct timespec t0 = { };
+	struct timespec t = { };
+	int i = 0;
+
+	get_current_time(&t0);
+	do {
+		for (i = 0; i < 100000; i++)
+			;
+		get_current_time(&t);
+	} while (timespec_diff_ns(&t0, &t) < (uint64_t)warmup * 1000000000);
+}
+
+static const char *yesno(int v)
+{
+	return (v ? "yes" : "no");
+}
+
+static double mb_per_sec(size_t size, double usec)
+{
+	return (1000000000 / usec) * ((double)size / (1024 * 1024));
+}
+
+static uint32_t get_curve_id(uint32_t width_bits)
+{
+	switch (width_bits) {
+	case ECC_CURVE_192:
+		return TEE_ECC_CURVE_NIST_P192;
+	case ECC_CURVE_224:
+		return TEE_ECC_CURVE_NIST_P224;
+	case ECC_CURVE_256:
+		return TEE_ECC_CURVE_NIST_P256;
+	case ECC_CURVE_384:
+		return TEE_ECC_CURVE_NIST_P384;
+	case ECC_CURVE_521:
+		return TEE_ECC_CURVE_NIST_P521;
+	default:
+		fprintf(stderr, "ECC curve is not supported!\n");
+	}
+
+	return TEE_CRYPTO_ELEMENT_NONE;
+}
+
+static int asym_perf_run_test(int mode, size_t size, uint32_t n,
+				     uint32_t l, int is_random, uint32_t warmup,
+				     uint32_t verbosity, uint32_t width_bits,
+				     uint32_t main_algo, uint32_t salt_len,
+				     uint32_t algo)
+{
+	TEEC_Operation op = TEEC_OPERATION_INITIALIZER;
+	uint32_t cmd = TA_CRYPTO_PERF_CMD_ASYM_PROCESS;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	uint8_t keygen_dh_p[DH_MAX_SIZE] = { };
+	uint8_t keygen_dh_g[DH_G_SIZE] = { };
+	TEE_Attribute params[4] = { };
+	size_t param_count = 0;
+	uint32_t curve_id = TEE_CRYPTO_ELEMENT_NONE;
+	struct statistics stats = { };
+	struct timespec ts = { };
+	struct timespec t0 = { };
+	struct timespec t1 = { };
+	uint32_t ret_origin = 0;
+	uint8_t *buf = NULL;
+	size_t blen = 0;
+	double sd = 0;
+	int n0 = n;
+
+	if (clock_getres(CLOCK_MONOTONIC, &ts) < 0) {
+		perror("clock_getres");
+		return -1;
+	}
+
+	verbose("random=%s, ", yesno(is_random == CRYPTO_USE_RANDOM));
+	verbose("inner loops=%u, loops=%u, warm-up=%u s\n", l, n, warmup);
+	vverbose("Clock resolution is %jd ns\n", (intmax_t)ts.tv_sec *
+		 1000000000 + ts.tv_nsec);
+
+	open_ta();
+
+	alloc_buffers(size, verbosity);
+	if (is_random == CRYPTO_USE_ZEROS)
+		memset((uint8_t *)in_shm.buffer, 0, size);
+	else
+		read_random(in_shm.buffer, size);
+	if (mode == MODE_DECRYPT && main_algo == ALGO_RSA)
+		/* Make sure the N > M */
+		((unsigned char *)(in_shm.buffer))[0] = 0x00;
```

Comment:
prefer with braces since the block spreads over 2 lines:
```c
	if (mode == MODE_DECRYPT && main_algo == ALGO_RSA) {
		/* Make sure the N > M */
		((unsigned char *)(in_shm.buffer))[0] = 0x00;
	}